node/primitives/src/lib.rs

```diff
@@ -56,6 +57,112 @@ impl Statement {
 	}
 }
 
+/// This helper trait ensures that we can encode Statement as CompactStatement,
+/// and anything as itself.
+pub trait EncodeAs<T> {
```

Comment:
There exists EncodeLike for this in scale codec.

Reply:
Part of the reason I wrote that trait was that I needed a blanket `impl<T: Encode> EncodeAs<T> for T`. No such thing is implemented for `EncodeLike`... but there is both a `impl<'_, T: Encode> EncodeLike<T> for &'_ T` and `impl<'_, T: Encode> EncodeLike<&'_ T> for T`. I guess the compiler is smart enough to chase that down, and decide that there's effectively an identity impl?

Reply:
@bkchr Reading the documentation for `EncodeLike`, it's not obvious to me how the conversion is actually performed. I.e. the example is 

```rust
fn encode_like(T: Encode, R: EncodeLike<T>>(data: R) -> Vec<u8> {
  data.encode() // valid encoding of T
}
```

How, in that example, does `R` actually get converted into `T` before encoding? It's a marker trait, so there's nowhere to perform a conversion within the trait implementation.

If it doesn't actually perform the conversion, but instead assert at the typesystem level that the encodings are equivalent, then I still need this `EncodeAs` trait. The whole point of encoding `Statement` as `CompactStatement` is to only actually encode a subset.

Reply:
Ahh okay, I thought these types encode to the same representation. As this is not the case, you approach is correct.